boxoffice/views/order.py

```diff
@@ -327,33 +327,112 @@ def format_assignee(line_item):
+    if not line_item.is_free:
+        order = line_item.order
+        if line_item.discount_policy:
+            pre_cancellation_order_amount = order.get_amounts(LINE_ITEM_STATUS.CONFIRMED).confirmed_amount
+            pre_cancellation_line_items = order.get_confirmed_line_items
+            line_item.cancel()
+            updated_order = update_order_on_line_item_cancellation(order, pre_cancellation_line_items, line_item)
+            post_cancellation_order_amount = updated_order.get_amounts(LINE_ITEM_STATUS.CONFIRMED).confirmed_amount
+            refund_amount = pre_cancellation_order_amount - post_cancellation_order_amount
+        else:
+            line_item.cancel()
+            refund_amount = line_item.final_amount
 
-    if line_item.final_amount > Decimal('0'):
         payment = OnlinePayment.query.filter_by(order=line_item.order, pg_payment_status=RAZORPAY_PAYMENT_STATUS.CAPTURED).one()
-        rp_resp = razorpay.refund_payment(payment.pg_paymentid, line_item.final_amount)
+        rp_resp = razorpay.refund_payment(payment.pg_paymentid, refund_amount)
         if rp_resp.status_code == 200:
-            line_item.cancel()
-            db.session.add(PaymentTransaction(order=line_item.order, transaction_type=TRANSACTION_TYPE.REFUND,
-                online_payment=payment, amount=line_item.final_amount, currency=CURRENCY.INR))
+            db.session.add(PaymentTransaction(order=order, transaction_type=TRANSACTION_TYPE.REFUND,
+                online_payment=payment, amount=refund_amount, currency=CURRENCY.INR))
             db.session.commit()
         else:
             raise PaymentGatewayError("Cancellation failed for order - {order} with the following details - {msg}".format(order=order.id,
                 msg=rp_resp.content), 424,
             'Refund failed. Please try again or write to us at {email}.'.format(email=line_item.order.organization.contact_email))
     else:
+        # free line item
+        refund_amount = Decimal(0)
         line_item.cancel()
         db.session.commit()
-    send_line_item_cancellation_mail.delay(line_item.id)
-    return make_response(jsonify(status='ok', result={'message': 'Ticket cancelled', 'cancelled_at': date_format(line_item.cancelled_at)}), 201)
+    return refund_amount
+
+
+@app.route('/line_item/<line_item_id>/cancel', methods=['POST'])
+@lastuser.requires_login
+@load_models(
+    (LineItem, {'id': 'line_item_id'}, 'line_item'),
+    permission='org_admin'
+    )
+def cancel_line_item(line_item):
+    if not line_item.is_cancellable():
+        return make_response(jsonify(status='error', error='non_cancellable', error_description='This ticket is not cancellable'), 403)
+
+    refund_amount = process_line_item_cancellation(line_item)
+    send_line_item_cancellation_mail.delay(line_item.id, refund_amount)
+    return make_response(jsonify(status='ok', result={'message': 'Ticket cancelled', 'cancelled_at': date_format(line_item.cancelled_at)}), 200)
```

Comment:
This JSON wrapper of status+result should be in a shared function. Maybe add an `api_result` function to Baseframe.